File: 73. Set Matrix Zeroes/setZeroes.py
from typing import List

class Solution:
  def setZeroes(self, matrix: List[List[int]]) -> None:
    """
    Do not return anything, modify matrix in-place instead.
    """
    ROWS = len(matrix)
    COLS = len(matrix[0])
    firstRowHasZero = False
    firstColHasZero = False
    
    for r in range(ROWS):
      if matrix[r][0] == 0:
        firstColHasZero = True
        break
    
    for c in range(COLS):
      if matrix[0][c] == 0:
        firstRowHasZero = True
        break 
      
    for r in range(ROWS):
      for c in range(COLS):
        if matrix[r][c] == 0:
          matrix[r][0] = 0
          matrix[0][c] = 0
     
    for r in range(1, ROWS):
      for c in range(1, COLS):
        if matrix[r][0] == 0 or matrix[0][c] == 0:
          matrix[r][c] = 0
         
    if firstColHasZero: 
      for r in range(ROWS):
        matrix[r][0] = 0
        
    if firstRowHasZero:
      for c in range(COLS):
        matrix[0][c] = 0
        
        
matrix = [
  [1, 1, 1],
  [1, 0, 1],
  [1, 1, 1]
]
solution = Solution()
print('setZeroes: ', solution.setZeroes(matrix))

File: 73. Set Matrix Zeroes/setZeroes.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
const setZeroes = (matrix) => {
  const ROWS = matrix.length;
  const COLS = matrix[0].length;
  let firstRowHasZero = false;
  let firstColHasZero = false;

  for (let r = 0; r < ROWS; r++) {
    if (matrix[r][0] === 0) {
      firstColHasZero = true;
      break;
    }
  }
  
  for (let c = 0; c < COLS; c++) {
    if (matrix[0][c] === 0) {
      firstRowHasZero = true;
      break;
    }
  }

  for (let r = 1; r < ROWS; r++) {
    for (let c = 1; c < COLS; c++) {
      if (matrix[r][c] === 0) {
        matrix[0][c] = 0;
        matrix[r][0] = 0;
      }
    }
  }

  for (let r = 1; r < ROWS; r++) {
    for (let c = 1; c < COLS; c++) {
      if (matrix[0][c] === 0 || matrix[r][0] === 0) {
        matrix[r][c] = 0;
      }
    }
  }

  if (firstColHasZero) {
    for (let r = 0; r < ROWS; r++) {
      matrix[r][0] = 0;
    }
  }

  if (firstRowHasZero) {
    for (let c = 0; c < COLS; c++) {
      matrix[0][c] = 0;
    }
  }
};


const matrix = [
  [1, 1, 1],
  [1, 0, 1],
  [1, 1, 1]
];
console.log('setZeroes: ', setZeroes(matrix));
